### DISCUSSION TITLE
Does "don't use symbols other than English letters" still make sense today?

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#6248] [Q&A]
Every teacher said so, but no one explained the reason to me.
I see in the history of C shar that he has used Unicode encoding since his birth.
That is to say, C Sharp has supported all national languages since the beginning.
I don't know the past of other languages, but I think they should be the same now.
We are a small company. Our colleagues are not good at English. They like to directly use the words of our country as variable names.
In the current document, there is still "do not use underscores, hyphens, or any other nonalphanumeric characters."
I'm confused. Is this agreement directly for large companies? Because they may have employees from different countries?
Or considering that they may raise questions in international forums? Or was it that people used to use English long ago, and now even if they can use other words, people are not used to it?
(Machinetranslation may not be able to express my meaning accurately. i 'm sorry)

**Comment by svick** [comment]
As a general guideline, I think it still makes sense, for several reasons:

1. Most of the libraries you will use are in English. So the choice isn't between "use English" and "use your local language", it's between "use English" and "use a mix of English and your local language". And I think the second choice is not very appealing.
2. Software development is quite international, so there is a decent chance you will want to hire a developer who doesn't speak your local language in the future. If you use the local language in your code, that makes it much harder.
3. Software developers mostly need to know English anyway, since most resources are English-only. So requiring English for code is not a huge burden. And if some developer doesn't know English well, pushing them to learn more will make them a better developer.

But even then, it's still a guideline, not a law. And the one you quoted is specifically for the public API of libraries, not things like local variables. And since you know your situation better than any guideline ever could, you can still decide for yourself if you're going to follow the guideline or ignore it.

(Note that I'm biased: while I'm not a native English speaker, I do use English almost exclusively in my professional life.)

**Comment by CyrusNajmabadi** [comment]
This is not a rule of c#. And I've seen many code bases that do not use English.

**Comment by BillWagner** [comment]
When I taught C# classes, often to non-native English speakers, I saw a mixture of uses.

- Some teams preferred all English.
- Some used English in the code, and all comments were in the team's native language.
- Some used English for keywords and all class and method names were in the team's native language.

There's no language requirement. The most important factor should be the team and others reading and maintaining the code.

**Comment by huoyaoyuan** [comment]
I'm not a native English speaker.

Generally, you should consider the properties of each codebase:
- Whom are the code for? This is the most important thing. For utilities dealing with famous Chinese website/resources, the documentation and comments will be all in Chinese of course.
- Mixture of different languages isn't that bad. It's more important to keep consistency and clarity. For example, you can always add comment in your native language.
- Consider productivity. Many Chinese developers don't like to use Chinese in code elements because it requires switching IME.
- Also be aware of technical issues when using non-ASCII characters. Do specify the encoding of files. UTF8-BOM seems to get best compatibility on Windows.
